File: TP/chapitre2/correction_tp_factory/src/creature/factory.py
from typing import Literal
from src.creature.implementation import Dragon, Phoenix, Griffin
from src.creature.abstract import Creature  


type CreatureType = Literal["dragon", "phoenix", "griffin"]


class CreatureFactory:

    _creatures: dict[CreatureType, type[Creature]] = {
        "dragon": Dragon,
        "phoenix": Phoenix,
        "griffin": Griffin
    }

    @staticmethod
    def create_creature(creature_type: CreatureType) -> Creature:

        if creature_type not in CreatureFactory._creatures:
            raise ValueError(f"Invalid creature type: {creature_type}")

        return CreatureFactory._creatures[creature_type]()
    
    @staticmethod
    def display_creatures() -> None:
        print("\nAvailable creatures:")
        print('==============')
        for creature_type in CreatureFactory._creatures:
            print(f" - {creature_type.capitalize()}")
        print('==============\n')

File: TP/chapitre2/correction_tp_factory_typescript/src/creature/factory.ts
import { Creature } from "./abstract";
import { Dragon, Griffin, Phoenix } from "./implementation";

export type CreatureType = "dragon" | "phoenix" | "griffin";

export class CreatureFactory {
  private static creatures: Record<CreatureType, () => Creature> = {
    dragon: () => new Dragon(),
    phoenix: () => new Phoenix(),
    griffin: () => new Griffin(),
  };

  public static display_creatures(): void {
    console.log("Available creatures:");
    Object.keys(CreatureFactory.creatures).forEach((creature) => {
      console.log(
        ` - ${creature.slice(0, 1).toUpperCase()}${creature.slice(1)}`
      );
    });
  }

  public static create_creature(creature_type: CreatureType): Creature {
    if (!(creature_type in CreatureFactory.creatures)) {
      throw new Error(`Invalid creature type: ${creature_type}`);
    }

    return CreatureFactory.creatures[creature_type]();
  }
}
